src/Storages/StorageIceberg.cpp

```diff
@@ -0,0 +1,261 @@
+#include "config.h"
+#if USE_AWS_S3
+
+#    include <Storages/StorageIceberg.h>
+#    include <Common/logger_useful.h>
+
+#    include <Columns/ColumnString.h>
+#    include <Columns/ColumnTuple.h>
+#    include <Columns/IColumn.h>
+
+#    include <Storages/StorageFactory.h>
+
+#    include <Formats/FormatFactory.h>
+
+#    include <fmt/format.h>
+
+#    include <Processors/Formats/Impl/AvroRowInputFormat.h>
+
+#    include <Poco/JSON/Array.h>
+#    include <Poco/JSON/Object.h>
+#    include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int S3_ERROR;
+    extern const int FILE_DOESNT_EXIST;
+    extern const int ILLEGAL_COLUMN;
+}
+
+template <typename Configuration, typename MetadataReadHelper>
+IcebergMetadataParser<Configuration, MetadataReadHelper>::IcebergMetadataParser(const Configuration & configuration_, ContextPtr context_)
+    : base_configuration(configuration_), context(context_)
+{
+}
+
+template <typename Configuration, typename MetadataReadHelper>
+std::vector<String> IcebergMetadataParser<Configuration, MetadataReadHelper>::getFiles() const
+{
+    auto metadata = getNewestMetaFile();
+    auto manifest_list = getManiFestList(metadata);
+
+    /// When table first created and does not have any data
+    if (manifest_list.empty())
+    {
+        return {};
+    }
+
+    auto manifest_files = getManifestFiles(manifest_list);
+    return getFilesForRead(manifest_files);
+}
+
+template <typename Configuration, typename MetadataReadHelper>
+String IcebergMetadataParser<Configuration, MetadataReadHelper>::getNewestMetaFile() const
+{
+    /// Iceberg stores all the metadata.json in metadata directory, and the
+    /// newest version has the max version name, so we should list all of them
+    /// then find the newest metadata.
+    static constexpr auto meta_file_suffix = ".json";
+    auto metadata_files = MetadataReadHelper::listFilesMatchSuffix(base_configuration, metadata_directory, meta_file_suffix);
+
+    if (metadata_files.empty())
+        throw Exception(
+            ErrorCodes::FILE_DOESNT_EXIST, "The metadata file for Iceberg table with path {} doesn't exist", base_configuration.url.key);
+
+    auto it = std::max_element(metadata_files.begin(), metadata_files.end());
```

Comment:
why do we need to take max?

Reply:
The iceberg table is multi version, each version corresponding to a `.json` metadata file, and the newest version has the largest version number.

Reply:
ok, let's put this explanation into comment in code?

Reply:
I have put comment in the beginning of the method.